[B004] 예약번호로 편지 수정 기능

kylekim2123 opened the pull request.

kylekim2123 committed:
feat(B004): 예약번호로 편지 수정 기능 추가

kylekim2123 committed:
refactor(B004): 새로운 편지 생성의 응답을 편지 전체 정보에서 예약번호로 변경

kylekim2123 committed:
feat(B005): 예약번호로 편지 삭제 기능 추가

kylekim2123 committed:
feat(B006): 예외 상태 및 메시지 정의

kylekim2123 committed:
refactor(B006): 서비스 로직 예외 종류 변경 및 ExceptionRule 적용

minnczi commented on src/main/java/com/snailmail/back/domain/Letter.java:
```diff
@@ -76,4 +78,21 @@ public void registerReservationKey() {
         this.reservationKey = MessageFormat.format("{0}{1}{2}", prefix, infix, suffix);
     }
+
+    public void updateSenderName(String senderName) {
```
builder pattern의 toBuilder 속성을 사용하면 update를 일일히 구현할 필요 없이 객체의 특정 속성만 바꿀수 있다고 하니, 이 방법도 고려해 보면 좋을거같아요!
e.g. @Builder(toBuilder = true)

참고: https://adjh54.tistory.com/96

kylekim2123 replied:
오 toBuilder가 있다는 걸 처음 알게 되었는데, 되게 편리한 기능인 것 같네요 ㅎㅎ 감사합니다.
그런데, 사실 별도의 update 메서드를 만들게 된 이유는 아래와 같은데요.
- 단순 setter보다는 register, update와 같은 용어를 통해 해당 메서드가 어떤 역할을 하는지 더 명시하기 위함
- password나 letterStatus와 같이 update 하지 않는 필드에 대해서는 update 기능을 제공하지 않도록 명시
- 특히나 reservationKey와 scheduledDate의 경우에는 값을 받아 특정 로직을 거친 후 변환된 값이 할당되기 때문에, 해당 로직을 update 메서드 안에 캡슐화하기 위함

이런 이유들 때문에 toBuilder를 하게 되면, 의도치 않은 필드까지 변경 가능성을 줄 수 있기 때문에 일단 해당 프로젝트에서는 update 메서드를 명시하는 것이 어떨까 싶습니다!

minnczi commented on src/main/java/com/snailmail/back/service/LetterService.java:
```diff
@@ -25,9 +27,29 @@ public LetterResponse createLetter(LetterCreateRequest request) {
+        letter.updateDurationAndScheduledDate(request.getDuration());
+        letter.updateContent(request.getContent());
+
+        return LetterReservationKeyResponse.from(letter.getReservationKey());
```
이 함수에서 실제 repository에 저장되는 부분이 누락된것 같습니다

kylekim2123 replied:
엔티티 객체의 값을 수정하면 영속성 컨텍스트의 1차 캐시에 반영이 되기 때문에, 트랜잭션이 끝나면서 자동으로 DB에 해당 변경 내용이 flush() 되고 commit()이 일어나게 됩니다. 그래서 별도로 save 해주지 않아도 DB에 반영됩니다!

kylekim2123 committed:
feat(B006): ControllerAdvice를 이용한 전역 예외처리 기능 추가
- 400 BAD REQUEST
- 404 NOT FOUND
- 405 METHOD NOT ALLOWED
- 500 INTERNAL SERVER ERROR

kylekim2123 committed:
refactor(B006): DTO 객체 기본 생성자의 접근 제어자를 protected로 변경

kylekim2123 committed:
feat: Bean Validation을 이용한 사용자 입력 예외처리 기능 추가

kylekim2123 committed:
style(B006): DTO 클래스 명 변경

kylekim2123 committed:
Merge pull request #2 from what-the-study/feature/B002

[B002] 새로운 편지 생성 기능

kylekim2123 committed:
Merge branch 'main' of github.com:what-the-study/snail-mail-back into feature/B004

kylekim2123 committed:
Merge pull request #5 from what-the-study/feature/B005

[B005] 예약번호로 편지 삭제 기능

kylekim2123 committed:
Merge pull request #6 from what-the-study/feature/B006

[B006] 각종 예외처리 추가

kylekim2123 committed:
Merge branch 'feature/B004' of github.com:what-the-study/snail-mail-back into feature/B004

kylekim2123 committed:
Merge branch 'feature/B005' of github.com:what-the-study/snail-mail-back into feature/B004
